File: zero-server/src/main/kotlin/com/zeroday/handler/handlers/NotificationHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.handler.reqStr
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import com.zeroday.service.NotificationService
import kotlinx.serialization.json.JsonObject

class NotificationHandler(
    private val notificationService: NotificationService
) : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.GET_NOTIFICATIONS,
        RequestTypes.MARK_NOTIFICATION_READ,
        RequestTypes.MARK_ALL_NOTIFICATIONS_READ
    )

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val playerId = ctx.authenticatedPlayerId()
            ?: return HandlerResult.Err(ServerError.fromMessage("Authentication required", ErrorCategory.AUTHENTICATION))
        return when (type) {
            RequestTypes.GET_NOTIFICATIONS -> list(ctx, playerId, payload)
            RequestTypes.MARK_NOTIFICATION_READ -> markRead(ctx, playerId, payload)
            RequestTypes.MARK_ALL_NOTIFICATIONS_READ -> markAllRead(ctx, playerId)
            else -> HandlerResult.Err(ServerError.fromMessage("Unsupported notification request: $type", ErrorCategory.VALIDATION))
        }
    }

    private suspend fun list(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val player = ctx.services.playerService.getPlayer(playerId)
            ?: return HandlerResult.Err(ServerError.fromMessage("Player not found", ErrorCategory.NOT_FOUND))
        val includeRead = payload["include_read"]?.let { (it as? kotlinx.serialization.json.JsonPrimitive)?.content } != "false"
        val limit = payload["limit"]?.let { (it as? kotlinx.serialization.json.JsonPrimitive)?.content?.toIntOrNull() } ?: 50
        val items = notificationService.list(player, includeRead, limit).map { it.toMap() }
        return HandlerResult.Ok(
            type = ResponseTypes.NOTIFICATIONS,
            payload = mapOf(
                "notifications" to items,
                "unread_count" to notificationService.unreadCount(player)
            )
        )
    }

    private suspend fun markRead(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val notifId = try { payload.reqStr("notification_id") } catch (_: Exception) {
            return HandlerResult.Err(ServerError.fromMessage("Notification ID required", ErrorCategory.VALIDATION))
        }
        val player = ctx.services.playerService.getPlayer(playerId)
            ?: return HandlerResult.Err(ServerError.fromMessage("Player not found", ErrorCategory.NOT_FOUND))
        val ok = notificationService.markRead(player, notifId)
        if (!ok) return HandlerResult.Err(ServerError.fromMessage("Unknown notification: $notifId", ErrorCategory.NOT_FOUND))
        return HandlerResult.Ok(
            type = ResponseTypes.NOTIFICATIONS,
            payload = mapOf("notification_id" to notifId, "unread_count" to notificationService.unreadCount(player))
        )
    }

    private suspend fun markAllRead(ctx: HandlerContext, playerId: String): HandlerResult {
        val player = ctx.services.playerService.getPlayer(playerId)
            ?: return HandlerResult.Err(ServerError.fromMessage("Player not found", ErrorCategory.NOT_FOUND))
        val n = notificationService.markAllRead(player)
        return HandlerResult.Ok(
            type = ResponseTypes.NOTIFICATIONS,
            payload = mapOf("marked_read" to n, "unread_count" to 0)
        )
    }
}

internal fun com.zeroday.model.Notification.toMap(): Map<String, Any?> = mapOf(
    "id" to id,
    "type" to type.name,
    "title" to title,
    "message" to message,
    "created_at" to createdAt,
    "read" to read,
    "data" to data
)

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/NotificationHandler.cs
using System.Text.Json;
using ZeroDayMMO.Server.Services;
using ZeroDayMMO.Shared;

namespace ZeroDayMMO.Server.Handlers;

public class NotificationHandler : IHandler
{
    private readonly INotificationService _notificationService;
    private readonly IPlayerService _playerService;

    public string MessageType => "notification";

    public NotificationHandler(INotificationService notificationService, IPlayerService playerService)
    {
        _notificationService = notificationService;
        _playerService = playerService;
    }

    public Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        if (payload is null)
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Missing payload" }));

        var action = payload.Value.Str("action");
        return action switch
        {
            "list" => HandleList(connectionId, payload.Value),
            "mark_read" => HandleMarkRead(connectionId, payload.Value),
            "mark_all_read" => HandleMarkAllRead(connectionId, payload.Value),
            _ => Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = $"Unknown notification action: {action}" }))
        };
    }

    private Task<IActionResult> HandleList(string connectionId, JsonElement payload)
    {
        var playerId = payload.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "playerId required" }));

        var player = _playerService.GetPlayer(playerId);
        if (player is null)
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Player not found" }));

        var includeRead = payload.TryGetProperty("include_read", out var ir)
            ? ir.ValueKind != JsonValueKind.False
            : true;
        var limit = payload.Int("limit") ?? 50;

        var notifications = _notificationService.GetNotifications(playerId, includeRead, limit);
        var unreadCount = _notificationService.GetUnreadCount(playerId);

        return Task.FromResult<IActionResult>(
            new MessageResult("notifications", connectionId, new
            {
                notifications,
                unread_count = unreadCount
            }));
    }

    private Task<IActionResult> HandleMarkRead(string connectionId, JsonElement payload)
    {
        var playerId = payload.Str("playerId");
        var notificationId = payload.Str("notification_id");

        if (string.IsNullOrEmpty(playerId) || string.IsNullOrEmpty(notificationId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "playerId and notification_id required" }));

        var player = _playerService.GetPlayer(playerId);
        if (player is null)
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Player not found" }));

        var ok = _notificationService.MarkRead(playerId, notificationId);
        if (!ok)
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = $"Unknown notification: {notificationId}" }));

        var unreadCount = _notificationService.GetUnreadCount(playerId);

        return Task.FromResult<IActionResult>(
            new MessageResult("notifications", connectionId, new
            {
                notification_id = notificationId,
                unread_count = unreadCount
            }));
    }

    private Task<IActionResult> HandleMarkAllRead(string connectionId, JsonElement payload)
    {
        var playerId = payload.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "playerId required" }));

        var player = _playerService.GetPlayer(playerId);
        if (player is null)
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Player not found" }));

        var notifications = _notificationService.GetNotifications(playerId, includeRead: false);
        var count = 0;
        foreach (var n in notifications)
        {
            if (_notificationService.MarkRead(playerId, n.Id))
                count++;
        }

        return Task.FromResult<IActionResult>(
            new MessageResult("notifications", connectionId, new
            {
                marked_read = count,
                unread_count = 0
            }));
    }
}
